spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -243,34 +243,59 @@ else if (this.mimeType.getType().equals("text") || this.mimeType.getSubtype().eq
 					if (this.provideHint) {
 						converted = ((AbstractMessageConverter) this.messageConverter).fromMessage(message, this.klazz,
 								this.mimeType);
+						if (converted == null && message.getHeaders().containsKey(MessageHeaders.CONTENT_TYPE)) {
+							converted = ((AbstractMessageConverter) this.messageConverter).fromMessage(
+									MessageConverterConfigurer.this.messageBuilderFactory.fromMessage(message)
+										.removeHeader(MessageHeaders.CONTENT_TYPE)
+										.build(), this.klazz, this.mimeType);
```

Comment:
Not sure what is that `messageBuilderFactory` configured for, but that seems for me it would be much efficient to use `MutableMessageBuilder` here any way.

That's really unfortunate that we have to rebuild message in this interceptor several time before it reaches the target `MH`...

I SI 5.0 we use `ConfigurableCompositeMessageConverter` for the `DefaultMessageHandlerMethodFactory` to enrich target `InvocableHandlerMethod` with proper argument conversion, including `Content-Type` negotiation.

I understand that this discussion isn't related to the PR, but just wanted to express the pain in my hurt each time I see message rebuild overhead...